62-tgyuuAn

tgyuuAn opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[우수 마을](https://www.acmicpc.net/problem/1949)

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
**2시간**

풀다가 못 풀고 ㅈㅈ쳤습니다...

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/116813010/0b71631a-8d1e-4b48-982e-a5503f6001f7)

<br><br><br><br><br><br>

### 처음 접근했던 잘못된 수도코드 🫡

<hr>

문제에서 주어진 보기와 같이 아래의 그림이 있을 때,

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/116813010/4b587221-d5f2-4966-a45f-54d08a522ebd)

**Root 노드** 부터 시작해서,

```python
DP[now] = max(DP[prev], DP[prev-prev] + 현재마을의인구값)
# 현재 위치의 최대 값 = 이전 마을의 최대 값, 더 이전 마을의 최대값+ 현재마을의 인구수
```

으로 DP값을 갱신한 뒤,

**Leaf 노드**에 도달했을 경우 해당 리프노드의 DP 값을 `answer` 에 더해주는 식으로 하려고 하였습니다.

<br><br><br>

그렇게 했을 경우, 아래와 같이 예시에서 주어진 인구 분포에서

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/116813010/3644efc5-be55-4558-bc71-8784d43003b2)

DP 마지막 값은 다음과 같이 갱신됩니다.



즉, **Leaf 노드**인 **5번 노드**에서 도달했을 때의 `DP[5]` 값은 7000이고,

**7번 노드**에 도달했을 때 `DP[7]`의 값은 10000이어서

`answer`가 17000으로 나오게 됩니다.

<br><br><br>

하지만 정답은 14000이죠.

17000은 나올 수가 없습니다.

왜냐하면 **5번 노드**에서 7000이 나올 조건은 **2번 노드가 선택되지 않아야만** 하지만,

**7번 노드**에서 10000이 나올 수 있는 조건은 **2번 노드가 선택되어야만 하기 때문**에 나올 수가 없는 경우였습니다.

<br><br><br><br><br><br>

그래서 골똘히 생각을 해보다가 결국 2시간만에 포기를 하고 레퍼런스를 참고하였습니다.

[참고한 레퍼런스](https://hyeo-noo.tistory.com/48)

<br><br><br>

레퍼런스를 보고 깨달은 사실은, 

**이전 마을을 선택했을 경우**와 **선택하지 않았을 경우의 DP 값을 각각 기록**해두면 위와 같은 상황이 방지 된다는 것이었다.

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/116813010/9cda990c-d455-4260-b3ab-3c1d1540418e)

<br><br><br>

추가적으로 레퍼런스들은 DFS를 사용하여 노드의 Leaf 노드부터 타고 올라오도록 구현하였다.

_(예를 들어 dfs에 1번노드 부터 시작하면 1번노드랑 연결되어있는 노드들의 leaf부터 탐색하여 마지막에 1번 DP값을 갱신하는 식이다.)_


## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->


<br><br><<br>

흠.............

**이전 마을을 선택했을 경우**와 **선택하지 않았을 경우의 DP 값을 각각 기록**해두는 것과 같은 생각들은 어떻게 하는 지 참 어렵다고 느꼈다.

한동안 바빠서 알고리즘 쉬었더니 DP적 감각이 많이 떨어진 것이 아닐까.

SeongHoonC reviewed (approved):
```kotlin

private lateinit var dp: Array<Array<Int>>
private lateinit var people: List<Int>
private lateinit var link: Array<MutableList<Int>>
private lateinit var visited: Array<Boolean>

fun main() {
    val br = System.`in`.bufferedReader()
    val n = br.readLine().toInt()
    people = listOf(0) + br.readLine().split(" ").map { it.toInt() }

    link = Array(n + 1) { mutableListOf() }
    visited = Array(n + 1) { false }

    repeat(n - 1) {
        val (first, second) = br.readLine().split(" ").map { it.toInt() }
        link[first].add(second)
        link[second].add(first)
    }

    dp = Array(n + 1) { arrayOf(0, people[it]) }
    dfs(1)
    println(max(dp[1][0], dp[1][1]))
}

private fun dfs(now: Int) {
    visited[now] = true
    for (next in link[now]) {
        if (!visited[next]) {
            dfs(next)
            dp[now][1] += dp[next][0]
            dp[now][0] += max(dp[next][1], dp[next][0])
        }
    }
}
```

음... 풀이 보고 이해하긴 했는데 쉽지 않았네요.

dfs 와 dp 를 같이 쓰는 것도 처음 봤는데
dp[now][0], dp[now][1] 을 분리해서 생각하는 것도 어려웠던 것 같아요..ㅎ

tgyuuAn commented:
> ```kotlin
> 
> 
> 
> private lateinit var dp: Array<Array<Int>>
> 
> private lateinit var people: List<Int>
> 
> private lateinit var link: Array<MutableList<Int>>
> 
> private lateinit var visited: Array<Boolean>
> 
> 
> 
> fun main() {
> 
>     val br = System.`in`.bufferedReader()
> 
>     val n = br.readLine().toInt()
> 
>     people = listOf(0) + br.readLine().split(" ").map { it.toInt() }
> 
> 
> 
>     link = Array(n + 1) { mutableListOf() }
> 
>     visited = Array(n + 1) { false }
> 
> 
> 
>     repeat(n - 1) {
> 
>         val (first, second) = br.readLine().split(" ").map { it.toInt() }
> 
>         link[first].add(second)
> 
>         link[second].add(first)
> 
>     }
> 
> 
> 
>     dp = Array(n + 1) { arrayOf(0, people[it]) }
> 
>     dfs(1)
> 
>     println(max(dp[1][0], dp[1][1]))
> 
> }
> 
> 
> 
> private fun dfs(now: Int) {
> 
>     visited[now] = true
> 
>     for (next in link[now]) {
> 
>         if (!visited[next]) {
> 
>             dfs(next)
> 
>             dp[now][1] += dp[next][0]
> 
>             dp[now][0] += max(dp[next][1], dp[next][0])
> 
>         }
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> 음... 풀이 보고 이해하긴 했는데 쉽지 않았네요.
> 
> 
> 
> dfs 와 dp 를 같이 쓰는 것도 처음 봤는데
> 
> dp[now][0], dp[now][1] 을 분리해서 생각하는 것도 어려웠던 것 같아요..ㅎ

진짜로요.... 다음에 DP 입대문제 추천!!!!!
